ddtrace/internal/_encoding.pyx

```diff
@@ -0,0 +1,357 @@
+                rawval = o
+                ret = msgpack_pack_bin(&self.pk, L)
+                if ret == 0:
+                    ret = msgpack_pack_raw_body(&self.pk, rawval, L)
+            elif PyUnicode_CheckExact(o):  #  if strict_types else PyUnicode_Check(o):
+                if self.encoding == NULL:
+                    ret = msgpack_pack_unicode(&self.pk, o, ITEM_LIMIT)
+                    if ret == -2:
+                        raise ValueError("unicode string is too large")
+                else:
+                    o = PyUnicode_AsEncodedString(o, self.encoding, self.unicode_errors)
+                    L = len(o)
+                    if L > ITEM_LIMIT:
+                        raise ValueError("unicode string is too large")
+                    ret = msgpack_pack_raw(&self.pk, L)
+                    if ret == 0:
+                        rawval = o
+                        ret = msgpack_pack_raw_body(&self.pk, rawval, L)
+            elif PyDict_CheckExact(o):
+                d = <dict>o
+                L = len(d)
+                if L > ITEM_LIMIT:
+                    raise ValueError("dict is too large")
+                ret = msgpack_pack_map(&self.pk, L)
+                if ret == 0:
+                    for k, v in d.items():
+                       ret = self._pack(k)
+                       if ret != 0: break
+                       ret = self._pack(v)
+                       if ret != 0: break
+            elif PyList_CheckExact(o):
+                # Expect a list of traces or a list of spans
+
+                L = len(o)
+                if L > ITEM_LIMIT:
+                    raise ValueError("list is too large")
+
+                ret = msgpack_pack_array(&self.pk, L)
+                if ret != 0:
+                    break
+
+                if L > 0 and PyList_CheckExact(o[0]):
+                    # List of traces
+                    for i in range(L):
+                        span = o[i]
+                        ret = self._pack(span)
+                        if ret != 0: break
+                else:
+                    # List of spans
+                    for i in range(L):
+                        ret = self._pack(o[i])
+                        if ret != 0: break
+
+            elif type(o) is Span:
+                L = 12
+                if L > ITEM_LIMIT:
+                    raise ValueError("list is too large")
+
+                has_span_type = <bint>(o.span_type is not None)
+                has_meta = <bint>(len(o.meta) > 0)
+                has_metrics = <bint>(len(o.metrics) > 0)
+
+                L = L - (1 - has_span_type) - (1 - has_meta) - (1 - has_metrics)
+
+                ret = msgpack_pack_map(&self.pk, L)
+
+                if ret == 0:
+                    ret = self._pack_bytes(<char *>b"trace_id")
+                    if ret != 0: return ret
+                    ret = self._pack(o.trace_id)
```

Comment:
same here, should we not use `self._pack` and just `msgpack_pack_long` directly?

`o.trace_id` should always be an int (maybe) either `int`/`float`/`str`/`bytes` only maybe

Reply:
There is some trickiness here because we'll be dependent on the Python implementation of numbers (see the checks that are done above). We could run into trickiness of `1` passes `PyInt_CheckExact` but `10000` passes `PyLong_CheckExact`. I opted not to inline it for now to maintain 100% compatibility with the previous implementation, but we should certainly note it for something to potentially optimize later.

Reply:
got it 👍